src/faebryk/core/cpp/src/pathfinder/bfs.cpp

```diff
@@ -0,0 +1,182 @@
+/* This file is part of the faebryk project
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "pathfinder/bfs.hpp"
+#include "perf.hpp"
+#include <deque>
+#include <sstream>
+
+std::string Edge::str() const {
+    std::stringstream ss;
+    ss << from->get_full_name(false) << "->" << to->get_full_name(false);
+    return ss.str();
+}
+
+std::string PathStackElement::str() /*const*/ {
+    std::stringstream ss;
+    if (up) {
+        ss << child_type.get_name() << "->" << parent_type.get_name() << "." << name;
+    } else {
+        ss << parent_type.get_name() << "." << name << "->" << child_type.get_name();
+    }
+    return ss.str();
+}
+
+bool UnresolvedStackElement::match(PathStackElement &other) {
+    return elem.parent_type == other.parent_type &&
+           elem.child_type == other.child_type && elem.name == other.name &&
+           elem.up != other.up;
+}
+
+std::string UnresolvedStackElement::str() /*const*/ {
+    std::stringstream ss;
+    ss << elem.str();
+    if (split) {
+        ss << " split";
+    }
+    return ss.str();
+}
+
+// BFSPath implementations
+BFSPath::BFSPath(/*const*/ GI_ref_weak path_head)
+  : Path(path_head)
+  , path_data(std::make_shared<PathData>()) {
+}
+
+BFSPath::BFSPath(const BFSPath &other)
+  : Path(other)
+  , path_data(std::make_shared<PathData>(*other.path_data))
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+}
+
+BFSPath::BFSPath(const BFSPath &other, /*const*/ GI_ref_weak new_head)
+  : Path(other)
+  , path_data(other.path_data)
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+    path.push_back(new_head);
+    assert(!other.filtered);
+}
+
+BFSPath::BFSPath(BFSPath &&other)
+  : Path(std::move(other))
+  , path_data(std::move(other.path_data))
+  , confidence(other.confidence)
+  , filtered(other.filtered)
+  , stop(other.stop) {
+}
+
+BFSPath BFSPath::operator+(/*const*/ GI_ref_weak gif) {
+    return BFSPath(*this, gif);
+}
+
+PathData &BFSPath::get_path_data_mut() {
+    if (!path_data.use_count() == 1) {
```

Comment:
See dangers on multithreading for future here: https://en.cppreference.com/w/cpp/memory/shared_ptr/use_count